filament/src/materials/flare/flare.mat

```diff
@@ -0,0 +1,162 @@
+            name : ghostSpacing,
+            precision: medium
+        },
+        {
+            type : float,
+            name : haloRadius,
+            precision: medium
+        },
+        {
+            type : float,
+            name : haloThickness,
+            precision: medium
+        }
+    ],
+    variables : [
+        vertex
+    ],
+    outputs : [
+        {
+            name : color,
+            target : color,
+            type : float4
+        }
+    ],
+    domain : postprocess,
+    depthWrite : false,
+    depthCulling : false
+}
+
+
+vertex {
+
+void dummy(){ }
+
+void postProcessVertex(inout PostProcessVertexInputs postProcess) {
+    postProcess.vertex.xy = postProcess.normalizedUV;
+#if defined(TARGET_METAL_ENVIRONMENT) || defined(TARGET_VULKAN_ENVIRONMENT)
+    // On metal/vulkan postProcess.normalizedUV has its origin at the left-top, but we need a
+    // uniform coordinate space for SAO. So, we flip the y coordinate so we have bottom-left UV
+    // coordinates across all backends.
+    postProcess.vertex.y = 1.0 - postProcess.vertex.y;
+#endif
+}
+
+}
+
+
+fragment {
+
+void dummy(){ }
+
+float smoothstep01(const float t) {
+    return t * t * (3.0 - 2.0 * t);
+}
+
+float ring(float x, const float c, const float r) {
+    return 1.0 - smoothstep01(min(abs(x - c) / r, 1.0));
+}
+
+vec3 sampleColor(const vec2 uv) {
+    // because we work on reduced resolution, we assume mediump uv is enough.
+    float level = materialParams.level;
+    float chromaticAberration = materialParams.chromaticAberration;
+    // make sure to avoid divide-by-zero in normalize
+    vec2 offset = normalize(0.5001 - uv) * chromaticAberration;
+    return vec3(
+        textureLod(materialParams_color, uv + offset, level).r,
+        textureLod(materialParams_color, uv,          level).g,
+        textureLod(materialParams_color, uv - offset, level).b
+    );
+}
+
+vec3 ghosts(const vec2 vertex_uv) {
+    // because we work on reduced resolution, we assume mediump uv is enough.
+    float ghostSpacing = materialParams.ghostSpacing;
+    float ghostCount = materialParams.ghostCount;
+    float threshold = materialParams.threshold.x;
+    vec2 uv = 1.0 - vertex_uv;
+    vec2 v = (0.5 - uv) * ghostSpacing;
+    vec3 color = vec3(0.0);
+    for (float i = 0.0; i < ghostCount; i += 1.0) {
+        vec2 suv = fract(uv + v * i);
+        vec3 s = sampleColor(suv);
+        s = clamp(s - threshold, 0.0, 1.0);
+        color += s;
+    }
+    return color;
+}
+
+vec3 halo(const vec2 uv) {
+    float haloRadius = materialParams.haloRadius;
+    float haloThickness = materialParams.haloThickness;
+    float threshold = materialParams.threshold.y;
+    float2 aspectRatio = materialParams.aspectRatio;
+
+    // make sure to avoid divide-by-zero in normalize
+    vec2 v = vec2(0.5001) - uv;
+    // handle aspect ratio
+    v.x *= aspectRatio.x;
+    v = normalize(v);
+    v.x *= aspectRatio.y;
+
+    vec3 s = sampleColor(uv + v * haloRadius);
```

Comment:
Can’t we skip this if chromatic aberration is turned off?

Reply:
We could make sampleColor do a single fetch of RGB instead of three r,g and b. We could add a bool in uniform to decide or have two shaders.
The cache should absorb all that though.